### ISSUE TITLE
Training doesn't work

### ISSUE THREAD

**Issue by aeyoa** [tucan9389/tf2-mobile-pose-estimation#19]
First of all, thank you for sharing the code 👍🏻I'm trying to get into ML for my project. I've installed all dependancies successfully, downloaded the dataset and started the training script with `python train.py`.

It seems to work fine. I'm getting output like this: 
```
Train for 100 steps
Epoch 1/500
2019-11-20 05:27:23.706954: I tensorflow/core/profiler/lib/profiler_session.cc:184] Profiler session started.
 99/100 [============================>.] - ETA: 2s - loss: 0.2444   
Epoch 00001: saving model to /Users/aeyoa/Documents/Sites/tf2-mpe/outputs/models/11200527_hg_lr0.0001.hdf5
100/100 [==============================] - 251s 3s/step - loss: 0.2422
```
After 500 epochs I'm getting loss of 0.0038 
```
Epoch 500/500
 99/100 [============================>.] - ETA: 1s - loss: 0.0038  
Epoch 00500: saving model to /Users/aeyoa/Documents/Sites/tf2-mpe/outputs/models/11200527_hg_lr0.0001.hdf5
100/100 [==============================] - 208s 2s/step - loss: 0.0038
```
But when I'm checking the progress in Tensorboard the training seems to be broken: heatmaps are changing randomly. Here are a few examples: 

**After 245 epochs**
![Screenshot 2019-11-20 at 19 30 52](https://user-images.githubusercontent.com/6937970/69445080-abe3a380-0d62-11ea-829c-73438dfb95bd.jpg)

**After 246 epochs — suddenly white**
![Screenshot 2019-11-20 at 19 37 03](https://user-images.githubusercontent.com/6937970/69445126-c6b61800-0d62-11ea-9518-784c08337e7a.jpg)

**After 306 epochs — still no results**
![Screenshot 2019-11-20 at 23 01 59](https://user-images.githubusercontent.com/6937970/69445323-1f85b080-0d63-11ea-9d00-f2a7a6f3d144.jpg)

After the fitting I've checked the model with `model.predict` and still didn't get any reasonable heatmap. 

Have you tried the latest code for training? Does is work? Do you have any idea why this isn't working? 

Thank you

**Comment by grin** [comment]
I'm seeing very similar broken prediction heatmaps after 70 epochs of training with loss 0.0046

**Comment by grin** [comment]
@tucan9389 I think I found some errors in the code. Here: https://github.com/tucan9389/tf2-mobile-pose-estimation/blob/5248196af945ec81ae1935fd7fbb482e0e3ccf72/hourglass_model.py#L106-L113

Based on my understanding of how the Hourglass network works, it shouldn't merge and average the intermediate heatmaps. Instead, it should just return:
```py
return hg_out, intermediate_heatmap_layers
```
And later, where you create the model, here: https://github.com/tucan9389/tf2-mobile-pose-estimation/blob/5248196af945ec81ae1935fd7fbb482e0e3ccf72/hourglass_model.py#L43-L45

The intermediate heatmaps should be set as outputs of the model, so it should look something like this:
```py
        hg_next_stage, l2s = self.build_network(inputs, trainable=trainable) 
       
        self.model = tf.keras.Model(inputs=inputs, outputs=l2s)
```

But for this to work, the data loader should also repeat the label set for each image `STAGE_NUM` times. So that each intermediate output would get its own labels. 

Or, am I missing something?

**Comment by grin** [comment]
So if anyone lands here: this repo is broken. I managed to get it working with the ^^^ above changes plus I had to replace the whole hourglass model. I used this implementation as a reference: https://github.com/yuanyuanli85/Stacked_Hourglass_Network_Keras/blob/eddf0ae15715a88d7859847cfff5f5092b260ae1/src/net/hg_blocks.py . It's implemented in Keras so it's easy to convert to TF2.

**Comment by aeyoa** [comment]
Thank you, @grin! I've already solved my problem using different approach. But it's interesting to get the clarification!

**Comment by tucan9389** [comment]
@grin 

Thank you for answering.

Currently, the master branch is not properly working. I'm trying to solve some trouble by making a `tmp` branch. The code of the `tmp` branch worked properly, and the conversion to the coreml (or tflite) model is being carried out now.
I will merge the `tmp` branch to the master soon.
